### PR TITLE
[Heartbeat] Drop only inheritable cap set and defer setuid to node fork

### LINKED ISSUE #33292: Heartbeat on Agent failed to start due to permission denied
This issue is created for SDH https://github.com/elastic/sdh-beats/issues/2585.
Confirmed for the beats version 8.3.3.

The hearbeat fails to start intermittently due to:
```
`Exiting: error setting default paths: failed to create data path /appl/cep/bin/data/Elastic/Agent/data/elastic-agent-0ffbed/run/84e76000-3e38-11ed-88e6-afd537e012ac/heartbeat--8.3.3: mkdir /appl/cep/bin/data/Elastic: permission denied[\\n`](file://n%60/)
```

The latest theory is that it might have something to do when and how the heartbeat seccomp filter is applied. Depending on the timing it looks like the hearbeat might try to create paths at the time when it is restricted from calling the ```mkdirat``` syscall.

Here is the latest comment from the SDH:

here is another piece of the puzzle, the file system audit when the problem happens. 
you can see that the process uid is root (0)
and the syscall ```mkdirat``` fails with EACCES on ```/home/amaus/work/elastic-agent-8.3.3-linux-x86_64444/data/elastic-agent-0ffbed/run```

```
type=UNKNOWN[1420] msg=audit(10/08/2022 18:14:32.425:3573) : subj_apparmor=unconfined 
type=PROCTITLE msg=audit(10/08/2022 18:14:32.425:3573) : proctitle=/home/amaus/work/elastic-agent-8.3.3-linux-x86_64444/data/elastic-agent-0ffbed/install/heartbeat-8.3.3-linux-x86_64/heartbeat -E 
type=PATH msg=audit(10/08/2022 18:14:32.425:3573) : item=1 name=/home/amaus/work/elastic-agent-8.3.3-linux-x86_64444/data/elastic-agent-0ffbed/run obj=? nametype=CREATE cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0 
type=UNKNOWN[1421] msg=audit(10/08/2022 18:14:32.425:3573) :

type=PATH msg=audit(10/08/2022 18:14:32.425:3573) : item=0 name=/home/amaus/work/elastic-agent-8.3.3-linux-x86_64444/data/elastic-agent-0ffbed/ inode=7607811 dev=08:05 mode=dir,775 ouid=amaus ogid=amaus rdev=00:00 obj=? nametype=PARENT cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0 
type=UNKNOWN[1421] msg=audit(10/08/2022 18:14:32.425:3573) :

type=CWD msg=audit(10/08/2022 18:14:32.425:3573) : cwd=/home/amaus/work/elastic-agent-8.3.3-linux-x86_64444/data/elastic-agent-0ffbed/install/heartbeat-8.3.3-linux-x86_64 
type=SYSCALL msg=audit(10/08/2022 18:14:32.425:3573) : arch=x86_64 syscall=mkdirat success=no exit=EACCES(Permission denied) a0=0xffffffffffffff9c a1=0xc00010e300 a2=0750 a3=0x7f1610425cc8 items=2 ppid=259214 pid=259245 auid=amaus uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=pts3 ses=3 comm=heartbeat exe=/home/amaus/work/elastic-agent-8.3.3-linux-x86_64444/data/elastic-agent-0ffbed/install/heartbeat-8.3.3-linux-x86_64/heartbeat subj=? key=elagent 
```

I also noticed that heartbeat defines the seccomp filter in:
https://github.com/elastic/beats/blob/main/heartbeat/security/policy_linux_amd64.go
that includes ```mkdirat```. 

I'm not familiar with the libbeat initialization order or the heartbeat or how the seccomp is utilized enough, but one possible concern that depending on the timing the paths initialization in the heartbeat could try to call ```mkdirat``` before this syscall was "allow-listed" with seccomp filter.

thoughts? maybe you or anybody who worked on the heartbeat closer might know about this.

@fearful-symmetry or @pierrehilbert I could dig further, or maybe you have somebody on the beats team who knows more about this area. I'll assign this ticket to you both for now, feel free to reassign as needed.

### PR DESCRIPTION
**Pull request by emilioalvap**
## What does this PR do?

Fixes #33292.

Only drop `inheritable` cap set and move `setuid/setgid` to node fork.

Still need to validate with inline and zip urls.

Process tree:
image

Caps:
image

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Build `heartbeat` locally and mount it into an `elastic-agent-complete` container.

### CODE REVIEW

File: heartbeat/security/security.go [hunk 1]
```diff
@@ -70,50 +70,33 @@ func changeUser(localUserName string) error {
...
-		return fmt.Errorf("error setting permitted setcap: %w", err)
-	}
-	err = newcaps.SetFlag(cap.Effective, true, cap.NET_RAW)
+	newcaps := cap.GetProc()
+
+	// Raise all permitted caps to effective
+	err := newcaps.Fill(cap.Effective, cap.Permitted)
 	if err != nil {
-		return fmt.Errorf("error setting effective setcap: %w", err)
+		return fmt.Errorf("error clearing inheritable cap set: %w", err)
```
Comment (emilioalvap) [inline heartbeat/security/security.go:93]:
```suggestion
		return fmt.Errorf("error raising effective cap set: %w", err)
```

File: x-pack/heartbeat/monitors/browser/synthexec/synthexec_linux.go [hunk 2]
```diff
@@ -1,26 +1,32 @@
 // Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 // or more contributor license agreements. Licensed under the Elastic License;
 // you may not use this file except in compliance with the Elastic License.
-//go:build linux || darwin
-// +build linux darwin
+//go:build linux
+// +build linux
```
Comment (emilioalvap) [inline x-pack/heartbeat/monitors/browser/synthexec/synthexec_linux.go:5]: Pdeathsig is Linux only, removing `darwin` from build tags to reflect that

**Comment by emilioalvap** [comment]
Tested the following scenarios:

- Heartbeat docker - root -> Monitors cannot be run as root err msg:

```zsh
io:job could not be initialized: script monitors cannot be run as root
```

- Heartbeat docker - non root -> No changes to process euid,  dropping limited caps.

```zsh
root@30728e612bb0:/usr/share/heartbeat# cat /proc/{8,23}/status | grep -E 'Name|Cap|Uid|Groups|Umask'
Name:   heartbeat
Uid:    1000    1000    1000    1000
Groups: 0 1000
CapInh: 0000000000000000
CapPrm: 0000000000002000
CapEff: 0000000000002000
CapBnd: 00000000a80425fb
CapAmb: 0000000000000000
Name:   node
Uid:    1000    1000    1000    1000
Groups: 0 1000
CapInh: 0000000000000000
CapPrm: 0000000000000000
CapEff: 0000000000000000
CapBnd: 00000000a80425fb
CapAmb: 0000000000000000
```

- Elastic agent docker - root -> All caps dropped, effective euid for node includes root (0):
root@dc536c6b5a5b:/usr/share/elastic-agent# cat /proc/{44,120}/status | grep -E 'Name|Cap|Uid|Groups|Umask'
```zsh
Name:   heartbeat
Umask:  0007
Uid:    0       0       0       0
Groups: 0
CapInh: 0000000000000000
CapPrm: 00000000a80425fb
CapEff: 00000000a80425fb
CapBnd: 00000000a80425fb
CapAmb: 0000000000000000
Name:   node
Umask:  0007
Uid:    1000    1000    1000    1000
Groups: 0
CapInh: 0000000000000000
CapPrm: 0000000000000000
CapEff: 0000000000000000
CapBnd: 00000000a80425fb
CapAmb: 0000000000000000
```

- Elastic agent docker - non-root -> No changes to process euid, dropping limited caps.
```zsh
root@ad2b16ed53b1:/usr/share/elastic-agent# cat /proc/{44,121}/status | grep -E 'Name|Cap|Uid|Groups|Umask'
Name:   heartbeat
Umask:  0007
Uid:    1000    1000    1000    1000
Groups: 0 1000
CapInh: 0000000000000000
CapPrm: 0000000000002080
CapEff: 0000000000002080
CapBnd: 00000000a80425fb
CapAmb: 0000000000000000
Name:   node
Umask:  0007
Uid:    1000    1000    1000    1000
Groups: 0 1000
CapInh: 0000000000000000
CapPrm: 0000000000000000
CapEff: 0000000000000000
CapBnd: 00000000a80425fb
CapAmb: 0000000000000000
```

cc @andrewvc @v, this should be green for review.

**Review by andrewvc — APPROVED** [review]
LGTM

**Pull request merged**